coderd/database/queries/notifications.sql

```diff
@@ -0,0 +1,144 @@
+-- of CODER_NOTIFICATIONS_LEASE_PERIOD. Once a message is delivered, its status is updated and the lease expires (set to NULL).
+-- If a message exceeds its lease, that implies the notifier did not shutdown cleanly, or the table update failed somehow,
+-- and the row will then be eligible to be dequeued by another notifier.
+--
+-- SKIP LOCKED is used to jump over locked rows. This prevents multiple notifiers from acquiring the same messages.
+-- See: https://www.postgresql.org/docs/9.5/sql-select.html#SQL-FOR-UPDATE-SHARE
+--
+-- name: AcquireNotificationMessages :many
+WITH acquired AS (
+    UPDATE
+        notification_messages
+            SET updated_at = NOW(),
+                status = 'leased'::notification_message_status,
+                status_reason = 'Leased by notifier ' || sqlc.arg('notifier_id')::int,
+                leased_until = NOW() + CONCAT(sqlc.arg('lease_seconds')::int, ' seconds')::interval
+            WHERE id IN (SELECT nm.id
+                         FROM notification_messages AS nm
+                                  LEFT JOIN notification_templates AS nt ON (nm.notification_template_id = nt.id)
+                         WHERE (
+                             (
+                                 -- message is in acquirable states
+                                 nm.status IN (
+                                               'pending'::notification_message_status,
+                                               'temporary_failure'::notification_message_status
+                                     )
+                                 )
+                                 -- or somehow the message was left in leased for longer than its lease period
+                                 OR (
+                                 nm.status = 'leased'::notification_message_status
+                                     AND nm.leased_until < NOW()
+                                 )
+                             )
+                           AND (
+                             -- exclude all messages which have exceeded the max attempts; these will be purged later
+                             nm.attempt_count IS NULL OR nm.attempt_count < sqlc.arg('max_attempt_count')::int
+                             )
+                           -- if set, do not retry until we've exceeded the wait time
+                           AND (
+                             CASE
+                                 WHEN nm.next_retry_after IS NOT NULL THEN nm.next_retry_after < NOW()
+                                 ELSE true
+                                 END
+                             )
+                         ORDER BY nm.created_at ASC
+                                  -- Ensure that multiple concurrent readers cannot retrieve the same rows
+                             FOR UPDATE OF nm
+                                 SKIP LOCKED
+                         LIMIT sqlc.arg('count'))
+            RETURNING id)
+SELECT
```

Comment:
One other idea: this query and also the `FetchNewMessageMetadata` query are complicated by the need to pull NotificationTemplate data.  But there is a small, finite number of NotificationTemplates, and we don't allow modifications...  so, we could just lazily load the NotificationTemplates and cache them in the notifier/manager.

That makes the queries simpler to understand (FetchNewMessageMetadata is probably not needed since you can just query the user), and likely faster to execute since Postgres doesn't need to do a join.

At some point down the line if we allow people to modify the templates at runtime, we'd need to invalidate the cache, but that seems pretty simple via the PubSub.

Reply:
> If that's too strange, I at least think it's clearer to return all the columns of notification_messages we want from the UPDATE, then do a single join to notification_templates, rather than returning the IDs and then doing two joins: back to itself and to notification_templates.

I decided to go for this option; thanks for saving a join 👍 
https://github.com/coder/coder/pull/13536/commits/c1a30103fbaff1d885f1f32e1c6a6459fe286bc0

> One other idea: this query and also the FetchNewMessageMetadata query are complicated by the need to pull NotificationTemplate data. But there is a small, finite number of NotificationTemplates, and we don't allow modifications... so, we could just lazily load the NotificationTemplates and cache them in the notifier/manager.

I agree, but for the sake of simplicity (and inertia, TBH) I think I'll keep this as-is for now and we can optimize/refactor it later.